nexus/pkg/server/stream.go

```diff
@@ -38,89 +39,134 @@ type Stream struct {
 	}
-	stream.wg.Add(1)
-	go stream.Start()
+	stream.Start()
 	return stream
 }
 
 // AddResponders adds the given responders to the stream's dispatcher.
 func (s *Stream) AddResponders(entries ...ResponderEntry) {
+	if s.responders == nil {
+		s.responders = make(map[string]Responder)
+	}
 	for _, entry := range entries {
-		s.dispatcher.AddResponder(entry)
+		responderId := entry.ID
+		if _, ok := s.responders[responderId]; !ok {
+			s.responders[responderId] = entry.Responder
+		} else {
+			s.logger.CaptureWarn("Responder already exists", "responder", responderId)
+		}
+	}
+}
+
+func (s *Stream) handleRespond(result *service.Result) {
+	responderId := result.GetControl().GetConnectionId()
+	s.logger.Debug("dispatch: got result", "result", result)
+	if responderId == "" {
+		s.logger.Debug("dispatch: got result with no connection id", "result", result)
+		return
+	}
+	response := &service.ServerResponse{
+		ServerResponseType: &service.ServerResponse_ResultCommunicate{
+			ResultCommunicate: result,
+		},
 	}
+	s.responders[responderId].Respond(response)
 }
 
 // Start starts the stream's handler, writer, sender, and dispatcher.
 // We use Stream's wait group to ensure that all of these components are cleanly
 // finalized and closed when the stream is closed in Stream.Close().
 func (s *Stream) Start() {
-	defer s.wg.Done()
-	s.logger.Info("created new stream", "id", s.settings.RunId)
+	// defer s.wg.Done()
+	// s.logger.Info("created new stream", "id", s.settings.RunId)
+
+	// TODO: fix input channel, either remove the defer state machine or make
+	// a pattern to handle multiple writers
 
 	// handle the client requests
-	s.wg.Add(1)
-	go func() {
-		s.handler.do()
-		s.wg.Done()
-	}()
+	s.handler = NewHandler(s.ctx, s.settings, s.logger)
+	handlerInChan := make(chan *service.Record, BufferSize)
+	handleChan, handlerResultChan := s.handler.do(handlerInChan)
 
 	// write the data to a transaction log
-	s.wg.Add(1)
-	go func() {
-		s.writer.do()
-		s.wg.Done()
-	}()
+	s.writer = NewWriter(s.ctx, s.settings, s.logger)
+	writerChan := s.writer.do(handleChan)
 
 	// send the data to the server
+	s.sender = NewSender(s.ctx, s.settings, s.logger)
+	senderResultChan, senderInChan := s.sender.do(writerChan)
+
+	s.logger.Debug("starting stream", "id", handlerResultChan)
+
+	// handle dispatching between components
 	s.wg.Add(1)
 	go func() {
-		s.sender.do()
+		for senderResultChan != nil || handlerResultChan != nil {
+			select {
+			case result, ok := <-senderResultChan:
+				if !ok {
+					senderResultChan = nil
+					continue
+				}
+				s.handleRespond(result)
+			case result, ok := <-handlerResultChan:
+				if !ok {
+					handlerResultChan = nil
+					continue
+				}
+				s.handleRespond(result)
+			}
+		}
+		s.logger.Debug("dispatch: finished")
 		s.wg.Done()
 	}()
 
-	// dispatch responses to the client
 	s.wg.Add(1)
+	streamInChan := s.inChan
 	go func() {
-		s.dispatcher.do()
+		for senderInChan != nil || streamInChan != nil {
+			select {
```

Comment:
I get the idea here, but one thing i dont like is that this complicates the basic datapath unnecessarily.
Now we introducing an additional channel copy to make appear more clean how to shutdown a channel.
An alternate approach is to add the complexity in the shutdown path where we are coordinating channel closes.
cuz the problem we are trying to solve is: "who closes X"... and this is the exact thing that you could solve by pushing a close message on the channel itself (what we effectively do with state machine logic)

in general i am fine with this approach, but i think we might revisit it once we build out more of the core communication paths and understand what a stream is managing besides the channel pipeline between handler, writer, sender

Reply:
I'll leave it as a todo in the code! This a a good point!